### ISSUE TITLE
fonts required for tests are unclear

### ISSUE THREAD

**Issue by kolewu** [lenmus/lomse#241]
While preparing the AUR-package of lomse I tried building in a clean chroot and was not able to fulfill the needs of some tests. Most of them I was able to resolve in adding the libertine font as a build dependency, but the last thing I was unable to find. The error message is this:

```
lomse_test_inlines_container_layouter.cpp(187):
    Failure in InlinesContainerLayouterTest / prepare_line_adds_bullet:
    is_equal_pos(lyt.my_get_available_space(), 4659.22f)
Test ChineseTextSplitterTest first_chunk skipped. Needed font not installed.
Test ChineseTextSplitterTest next_chunk skipped. Needed font not installed.
```

My best guess is, that the number is only correct for a special font ...

Are the included fonts not enough for the tests or is the setup wrong that configures them? How could I find out?

Out of the clean chroot, the tests run successfully.

**Comment by cecilios** [comment]
The test that fails (and some others) depend on font "Liberation serif".

The excluded Chinese test depend on font "NotoSansCJK" ("Noto Sans CJK SC").

And yes, those test are very fragile as they depend on a particular font and worse, comparing for a specific size with decimals is not a good idea. These tests should be fixed but I always have more urgent tasks!

Are you preparing a package for the lomse library? Or is it for the LenMus application?

**Comment by kolewu** [comment]
Yes, it's the AUR package of liblomse: https://aur.archlinux.org/packages/liblomse/

With the additional build time dependency of *ttf-liberation* the test now runs successful. But since the font is already included, there seems to be something wrong with the configuration of the testapp.

I noticed something else: The test is run twice, once when I call `make` and second when I explicitly call it.

**Comment by kolewu** [comment]
The font path is set correct, as I found out with this addition to the runner:
```diff
--- lomse_the_test_runner__orig.cpp	2020-05-23 13:10:42.830454286 +0200
+++ lomse_the_test_runner.cpp	2020-05-23 13:11:03.483110714 +0200
@@ -124,7 +124,8 @@
          << ". Library tests runner." << endl << endl;
 
     cout << "Lomse build date: " << LibraryScope::get_build_date() << endl;
-    cout << "Path for tests scores: '" << TESTLIB_SCORES_PATH << "'" << endl << endl;
+    cout << "Path for test scores: '" << TESTLIB_SCORES_PATH << "'" << endl;
+    cout << "Path for test fonts : '" << TESTLIB_FONTS_PATH << "'" << endl << endl;
 
     bool verbose = argc > 1 && (!strcmp(argv[1], "--verbose") || !strcmp(argv[1], "-v"));
     int nextArg = verbose ? 2 : 1;
```
so I guess there's something wrong with the loading of fonts that are not installed.

**Comment by kolewu** [comment]
Found the problem with doubled test: If I want to run the tests in an extra step, I have to exclude them from the build step with the cmake option `-DLOMSE_RUN_TESTS=no`.

**Comment by cecilios** [comment]
### test fonts

You have found a bug! When I changed the program to use fontconfig I forget about the TESTLIB_FONTS_PATH and, currently, it is ignored and the tests are using only installed fonts. I have to fix this.


### tests run twice

Current CMakeLists.txt file runs automatically the tests when it is requested to build them, so you don't need to explicitly run the tests. You can prevent this with -DLOMSE_RUN_TESTS=OFF as you found.

My knowledge about packaging is very small so in general, is running the tests automatically a problem for package maintainers? Should the default option not to run the tests automatically after building?


### Another issue

lomse has an install dependency on Bravura.otf font, that is not listed in https://aur.archlinux.org/packages/liblomse/ and, probably, this font is not installed by default in ArchLinux. This font is used for rendering the musical symbols.
Are you including the font in the package?

If not, I don't know if Bravura.otf is available in ArchLinux. As I have not found a Bravura font package for Debian and probably it is not yet available in other distros, I've created a .deb package at https://github.com/lenmus/bravura.and the Bravura.otf font is also in that repo. In any case, the official repo for this font is at https://github.com/steinbergmedia/bravura

**Comment by kolewu** [comment]
### tests run twice

It's not a problem, when the test is automatically run while building. But in Archlinux packaging, there is a separate step for checking. Since it's possible to leave the test out of building and call it separately, as far as Arch's packaging is concerned, it's fine.

### Bravura

Since Bravura is not tested nor listed as dependency for Lomse and it is included in the package, I have not added the dependency so far.

There is a small typo in `src/platform/lomse_linux.cpp` on [line 173](https://github.com/lenmus/lomse/blob/master/src/platform/lomse_linux.cpp#L173): The fontname is missing the leading `B`.

### fontconfig

Lenmus is working without installed Bravura font (and it is part of the package, too) or do you use another mechanism there to load the included font?

**Comment by cecilios** [comment]
> There is a small typo in src/platform/lomse_linux.cpp on line 173: The fontname is missing the leading B.

Thanks for pointing, but this is on purpose, just in case in some distro they decide to use lower case for the font name. I have added a comment to inform code reviewers and prevent false issues with this. 

> Lenmus is working without installed Bravura font

As most Linux distros do not have the Bravura font, lomse is flexible: it looks for this font in system installed fonts, via fontconfig, and if not fonud, looks for it at LOMSE_FONTS_PATH.  In this way, I do not force to create an install dependency form Bravura font.

For LenMus, the default option for building and installing is to include the Bravura font in the package and informing lomse about the path, thus not creating an install dependency from Bravura. But for very strict distros, such as Debian, that do not accept fonts in non-font packages, there is the option of not including the font in the package and create an install dependency.

In short, if liblomse package includes the Bravura font there is no install dependency from that font. Otherwise, it is necessary to define that dependecy.

**Comment by grantsloman** [comment]
Hi there,

Related question (possibly) - In order to get tutorial-1-wx working (on Mac OSX) I've needed to add the following in order for it to be able to find the Bravura font. Is this what you'd expect? Thanks :)

`     //initialize the library with these values
     m_lomse.init_library(pixel_format,resolution, reverse_y_axis);
+    
+    m_lomse.set_default_fonts_path("../../../fonts/");
`

**Comment by cecilios** [comment]
I have no experience with OSX and never compiled lomse there. Depending on how lomse is configured, it will use either `lomse/src/platform/lomse_linux.cpp` or `lomse/src/platform/lomse_other.cpp`. Both should be usable in OSX.

When using `lomse_other.cpp` the responsibility for managing fonts is on the user application. In this case your application has to invoke  `m_lomse.set_default_fonts_path("....")` and specify the location of needed fonts.

When using `lomse_linux.cpp`, lomse will require linking to fontconfig library. And this library will take care of managing fonts. Invoking `m_lomse.set_default_fonts_path("....")` in your application is not necessary but does not harm.

**Comment by grantsloman** [comment]
Thanks Cecilio.

The build is using `lomse_linux.cpp` and linking to fontconfig.

I've found a config file at `/usr/local/etc/fonts/fonts.conf` which specifies a directory of `/usr/share/fonts` (as well as the Mac system dirs) however, /usr/share/fonts doesn't exist.

I'm wondering if there needs to be an OSX-specific step for install in CMakeLists.txt to copy the fonts to /usr/share/fonts ? However, there's no problem with me calling `m_lomse.set_default_fonts_path("....")` as a workaround.

**Comment by cecilios** [comment]
> however, /usr/share/fonts doesn't exist.

fontconfig is a Linux package, that also works in Windows and Mac, but as it is not normally used Mac, probably it is not well maintained in OSX. Sorry, but I never used OSX so know nothing about OSX fonts paths.

When using `lomse_linux.cpp` and font config I think that the content of `/usr/share/fonts` will be irrelevant as fonts will be searched by fontconfig and if it is not properly configured probably will not find the fonts. So perhaps the first approach I suggets is trying to build using `lomse_other.cpp` and your application would supply the fonts.

I see (in file "lenmus/lomse"trunk/add-sources.cmake", line 228) that cmake is currently configured for using `lomse_linux.cpp` and  thus `fontconfig` when building for OSX. If you change that line to `if(UNIX AND NOT APPLE)` it will use lomse_other.cpp.

as to your question of installing fonts, I do not remeber the details at this moment but I think that by default the Bravura music font is always installed. Nevertheless, you can force to install the font by building with:
```
    cmake -DLOMSE_INSTALL_BRAVURA_FONT=ON [...]
```
Please, see options in the comments at begining of file CMakeLists.txt

**Comment by grantsloman** [comment]
Thanks, yes, using `lomse_other.cpp` and setting the fonts path manually for OSX sounds sensible. That would also remove the dependency on fontconfig for OSX builds.

**Comment by grantsloman** [comment]
Hi Cecilio, (I've just got back to this) I've made the change to use lomse_other.cpp instead of lomse_linux.cpp but I now have a problem where 4 Chinese text splitter unit tests fail, they're failing to find the required font.

Adding some debug in lomse_other.cpp to `FontSelector::find_font()` shows the name parameter being passed is "Noto Sans CJK SC" and the language parameter is "en".

The code in find_font() suggests that the required Chinese font file will only be selected if the language parameter is "zh_CN"...

`    //Chinese fonts`
`    else if (language == "zh_CN")`
`    {`
`        fullpath += "wqy-zenhei.ttc";`
`    }`

I'm not too sure what the correct approach to fixing this is, so your advice would be appreciated!

**Comment by cecilios** [comment]
Hi Grant,

> I'm not too sure what the correct approach to fixing this is, so your advice would be appreciated!

If all you need is fixing the tests, then in:
https://github.com/lenmus/lomse/blob/5fd2f47c94f8ce79502851bd8dcc239238792827/src/tests/lomse_test_text_splitter.cpp#L292-L299

please replace line 298:
```
    return pPara->add_text_item(data, pStyle);
````

By
```
    ImoTextItem* pText = pPara->add_text_item(data, pStyle);
    pText->set_language("zh_CN");
    return pText;
```

I will include this change in next PR.

But if you would like to improve font management in general, then everything is open to be changed. File `lomse_other.cpp` is just a very simple fallback solution that can be greately improved. 

Regards.

**Comment by cecilios** [comment]
Closing this old issue. For completness and future reference, just some additional information:

The excluded Chinese tests depend on font "NotoSansCJK" ("Noto Sans CJK SC").

Install in Linux:
    sudo apt-get install -y fonts-noto-cjk

Install in macOS
    brew tap homebrew/cask-fonts
    brew install --cask font-noto-sans-cjk
